### PR TITLE
Adding the scm_track_submodules option for project

### PR DESCRIPTION
**Pull request by mazhead**
##### SUMMARY
Adding the scm_track_submodules option which should fix the related #7846

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
 - API
 - UI

##### AWX VERSION
```
awx: 17.0.1
```

##### ADDITIONAL INFORMATION
This option will add the track_submodules option which is described in the ansible git module: https://docs.ansible.com/ansible/latest/collections/ansible/builtin/git_module.html

> if yes, submodules will track the latest commit on their master branch (or other branch specified in .gitmodules). If no, submodules will be kept at the revision specified by the main project. This is equivalent to specifying the --remote flag to git submodule update.

### CODE REVIEW

**Comment by one-t** [comment]
link #7846

**Comment by mazhead** [comment]
Hello,
Is there anything outstanding on this pull request?

**Comment by mazhead** [comment]
Is there anything outstanding on this?

**Comment by mazhead** [comment]
Are you guys even considering this? I will wait for your feedback and then rebase **again** if required....

**Comment by shanemcd** [comment]
@mazhead Apologies for not seeing this sooner. If you rebase one more time we'll get this reviewed for you.

**Comment by mazhead** [comment]
Seems like you have some issues on your test infra:
`static | Error response from daemon: unauthorized: You don't have the needed permissions to perform this operation, and you may have invalid credentials. To authenticate your request, follow the steps in: https://cloud.google.com/container-registry/docs/advanced-authentication`

**Comment by mazhead** [comment]
recheck

**Comment by shanemcd** [comment]
@mazhead That's not the actual error:

```
2021-04-12 14:30:42.146577 | static | awx-manage check_migrations --dry-run --check  -n 'missing_migration_file'
2021-04-12 14:30:45.200296 | static | Traceback (most recent call last):
2021-04-12 14:30:45.200447 | static |   File "/usr/local/bin/awx-manage", line 9, in <module>
2021-04-12 14:30:45.200538 | static |     load_entry_point('awx', 'console_scripts', 'awx-manage')()
2021-04-12 14:30:45.200618 | static |   File "/awx_devel/awx/__init__.py", line 164, in manage
2021-04-12 14:30:45.200680 | static |     execute_from_command_line(sys.argv)
2021-04-12 14:30:45.200873 | static |   File "/var/lib/awx/venv/awx/lib64/python3.8/site-packages/django/core/management/__init__.py", line 381, in execute_from_command_line
2021-04-12 14:30:45.200921 | static |     utility.execute()
2021-04-12 14:30:45.201069 | static |   File "/var/lib/awx/venv/awx/lib64/python3.8/site-packages/django/core/management/__init__.py", line 375, in execute
2021-04-12 14:30:45.201162 | static |     self.fetch_command(subcommand).run_from_argv(self.argv)
2021-04-12 14:30:45.201313 | static |   File "/var/lib/awx/venv/awx/lib64/python3.8/site-packages/django/core/management/base.py", line 323, in run_from_argv
2021-04-12 14:30:45.201372 | static |     self.execute(*args, **cmd_options)
2021-04-12 14:30:45.201488 | static |   File "/awx_devel/awx/main/management/commands/check_migrations.py", line 13, in execute
2021-04-12 14:30:45.201564 | static |     return MakeMigrations().execute(*args, **options)
2021-04-12 14:30:45.201708 | static |   File "/var/lib/awx/venv/awx/lib64/python3.8/site-packages/django/core/management/base.py", line 364, in execute
2021-04-12 14:30:45.201771 | static |     output = self.handle(*args, **options)
2021-04-12 14:30:45.201931 | static |   File "/var/lib/awx/venv/awx/lib64/python3.8/site-packages/django/core/management/base.py", line 83, in wrapped
2021-04-12 14:30:45.201993 | static |     res = handle_func(*args, **kwargs)
2021-04-12 14:30:45.202161 | static |   File "/var/lib/awx/venv/awx/lib64/python3.8/site-packages/django/core/management/commands/makemigrations.py", line 87, in handle
2021-04-12 14:30:45.202247 | static |     loader = MigrationLoader(None, ignore_no_migrations=True)
2021-04-12 14:30:45.202390 | static |   File "/var/lib/awx/venv/awx/lib64/python3.8/site-packages/django/db/migrations/loader.py", line 49, in __init__
2021-04-12 14:30:45.202430 | static |     self.build_graph()
2021-04-12 14:30:45.202577 | static |   File "/var/lib/awx/venv/awx/lib64/python3.8/site-packages/django/db/migrations/loader.py", line 206, in build_graph
2021-04-12 14:30:45.202616 | static |     self.load_disk()
2021-04-12 14:30:45.202762 | static |   File "/var/lib/awx/venv/awx/lib64/python3.8/site-packages/django/db/migrations/loader.py", line 108, in load_disk
2021-04-12 14:30:45.202852 | static |     migration_module = import_module(migration_path)
2021-04-12 14:30:45.202960 | static |   File "/usr/lib64/python3.8/importlib/__init__.py", line 127, in import_module
2021-04-12 14:30:45.203055 | static |     return _bootstrap._gcd_import(name[level:], package, level)
2021-04-12 14:30:45.203150 | static |   File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
2021-04-12 14:30:45.203241 | static |   File "<frozen importlib._bootstrap>", line 991, in _find_and_load
2021-04-12 14:30:45.203343 | static |   File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
2021-04-12 14:30:45.203435 | static |   File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
2021-04-12 14:30:45.203533 | static |   File "<frozen importlib._bootstrap_external>", line 779, in exec_module
2021-04-12 14:30:45.203628 | static |   File "<frozen importlib._bootstrap_external>", line 916, in get_code
2021-04-12 14:30:45.203730 | static |   File "<frozen importlib._bootstrap_external>", line 846, in source_to_code
2021-04-12 14:30:45.203846 | static |   File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
2021-04-12 14:30:45.203952 | static |   File "/awx_devel/awx/main/migrations/0136_scm_track_submodules.py", line 9
2021-04-12 14:30:45.204041 | static |     <<<<<<< HEAD:awx/main/migrations/0136_scm_track_submodules.py
2021-04-12 14:30:45.204063 | static |     ^
2021-04-12 14:30:45.204109 | static | SyntaxError: invalid syntax
2021-04-12 14:30:45.638850 | static | make: *** [Makefile:305: test] Error 1
2021-04-12 14:30:46.133946 | static | ERROR
2021-04-12 14:30:46.134540 | static | {
2021-04-12 14:30:46.134614 | static |   "delta": "0:05:57.455967",
2021-04-12 14:30:46.134655 | static |   "end": "2021-04-12 14:30:45.979342",
2021-04-12 14:30:46.134693 | static |   "msg": "non-zero return code",
2021-04-12 14:30:46.134728 | static |   "rc": 2,
2021-04-12 14:30:46.134764 | static |   "start": "2021-04-12 14:24:48.523375"
2021-04-12 14:30:46.134799 | static | }
2021-04-12 14:30:46.199568 | 
```

**Comment by shanemcd** [comment]
And for linter:

```
2021-04-12 14:25:20.173898 | static | error: cannot format /awx_devel/awx/main/migrations/0136_scm_track_submodules.py: Cannot parse: 9:0: <<<<<<< HEAD:awx/main/migrations/0136_scm_track_submodules.py
2021-04-12 14:25:34.906779 | static | would reformat /awx_devel/awx_collection/plugins/modules/tower_project.py
2021-04-12 14:25:38.273152 | static | Oh no! 💥 💔 💥
2021-04-12 14:25:38.273322 | static | 1 file would be reformatted, 813 files would be left unchanged, 1 file would fail to reformat.
2021-04-12 14:25:38.322354 | static | make: *** [Makefile:276: black] Error 123
2021-04-12 14:25:38.322951 | static | ERROR: InvocationError for command /usr/bin/make black (exited with code 2)
2021-04-12 14:25:38.324326 | static | ___________________________________ summary ____________________________________
2021-04-12 14:25:38.324418 | static | ERROR:   linters: commands failed
2021-04-12 14:25:38.819190 | static | ERROR
2021-04-12 14:25:38.819388 | static | {
2021-04-12 14:25:38.819437 | static |   "delta": "0:00:48.161367",
2021-04-12 14:25:38.819468 | static |   "end": "2021-04-12 14:25:38.627574",
2021-04-12 14:25:38.819494 | static |   "msg": "non-zero return code",
2021-04-12 14:25:38.819521 | static |   "rc": 1,
2021-04-12 14:25:38.819548 | static |   "start": "2021-04-12 14:24:50.466207"
2021-04-12 14:25:38.819574 | static | }
```

Once you fix the git error, just run `make black` to auto-format the files.

**Comment by mazhead** [comment]
@shanemcd  on it, thanks

**Comment by mazhead** [comment]
@shanemcd  -> should be done

**Comment by shanemcd** [comment]
@mazhead When I try to run this locally, I get the following error when the app boots:

```
awx_1_1     | Operations to perform:
awx_1_1     |   Apply all migrations: auth, conf, contenttypes, main, oauth2_provider, sessions, sites, social_django, sso, taggit
awx_1_1     | Running migrations:
postgres_1  | 2021-04-12 15:17:15.276 UTC [33] ERROR:  column "scm_track_submodules" contains null values
postgres_1  | 2021-04-12 15:17:15.276 UTC [33] STATEMENT:  ALTER TABLE "main_project" ADD COLUMN "scm_track_submodules" boolean NOT NULL
awx_1_1     |   Applying main.0136_scm_track_submodules...Traceback (most recent call last):
awx_1_1     |   File "/var/lib/awx/venv/awx/lib64/python3.8/site-packages/django/db/backends/utils.py", line 84, in _execute
awx_1_1     |     return self.cursor.execute(sql, params)
awx_1_1     | psycopg2.errors.NotNullViolation: column "scm_track_submodules" contains null values
```

**Comment by mazhead** [comment]
@shanemcd -> strange I did not experience this on a clean db. To be sure I have added a default in the migration. Cleaned my local env and re-executed:

```
awx_1_1     |   Applying main.0134_unifiedjob_ansible_version... OK
awx_1_1     |   Applying main.0135_schedule_sort_fallback_to_id... OK
awx_1_1     |   Applying main.0136_scm_track_submodules... OK
awx_1_1     |   Applying main.0137_auto_20210412_1621... OK
awx_1_1     |   Applying oauth2_provider.0002_08_updates... OK
awx_1_1     |   Applying oauth2_provider.0003_auto_20160316_1503... OK
```

Also double checked the migrations in the container:
```
bash-4.4$ awx-manage migrate
Operations to perform:
  Apply all migrations: auth, conf, contenttypes, main, oauth2_provider, sessions, sites, social_django, sso, taggit
Running migrations:
  No migrations to apply.
bash-4.4$ ./manage.py makemigrations
No changes detected
```

So hopefully it is "fixed", please check on your side with the latest commit.

File: awx/ui_next/src/screens/Project/shared/ProjectSubForms/SharedFields.jsx
Comment (shanemcd) [inline awx/ui_next/src/screens/Project/shared/ProjectSubForms/SharedFields.jsx]:
```suggestion
              t`If yes, submodules will track the latest commit on
```

**Comment by shanemcd** [comment]
No longer seeing the error, thanks for fixing!

**Review by shanemcd — APPROVED** [review]
I'll wait for another person or two to provide feedback, but this looks great to me. Thanks a bunch for the contribution!

File: awx/main/migrations/0136_scm_track_submodules.py
Comment (AlanCoding) [inline awx/main/migrations/0136_scm_track_submodules.py:16]: If the addition of `AlterField` operations are merely due to a change in help text, then I'd prefer to condense then into just 2 `AddField` operations before merging.

**Comment by AlanCoding** [comment]
My general feelings here are :+1: 

This is adding a project field to match with the options for the `git` module and nothing else. I totally support that.

This is only valid for git projects. I argue that we should reject a non-blank value for non-git projects, as we do here for the refspec, a git-specific concept.

https://github.com/ansible/awx/blob/b82318161c5ea737621106da139da285c49b10dc/awx/api/serializers.py#L1385-L1386

So that's a minor request for change that I'm making.

I want @wenottingham to look at this. This will need docs followup by @tvo318 eventually because it's very user-facing.

**Review by AlanCoding — APPROVED** [review]

File: awx/main/migrations/0136_scm_track_submodules.py
Comment (mazhead) [reply awx/main/migrations/0136_scm_track_submodules.py:16]: Modified in latest commit to be just 2 AddField

**Comment by mazhead** [comment]
Thanks @AlanCoding for the review, hope I have addressed all the points in last commit.

**Comment by mazhead** [comment]
Is anything more needed on this one from my side?

File: awx/ui_next/src/screens/Project/shared/ProjectSubForms/SharedFields.jsx
Comment (AlexSCorey) [inline awx/ui_next/src/screens/Project/shared/ProjectSubForms/SharedFields.jsx]: I would change this string just to remove the `If yes,` to be in keeping with other checkbox tooltips.

**Review by AlexSCorey — CHANGES_REQUESTED** [review]
This is really good work.  Just 1 small nitpicky change.

File: awx/ui_next/src/screens/Project/shared/ProjectSubForms/SharedFields.jsx
Comment (mazhead) [reply awx/ui_next/src/screens/Project/shared/ProjectSubForms/SharedFields.jsx]: Modified as requested

**Comment by mazhead** [comment]
@AlexSCorey -> done

**Review by AlexSCorey — APPROVED** [review]
Sweet, thanks!  LGTM

**Comment by mazhead** [comment]
So we still miss @jakemcdermott or anyone else?

**Pull request merged**
